modules/channel-messenger/src/backend/messenger.ts

```diff
@@ -0,0 +1,253 @@
+import axios, { AxiosInstance } from 'axios'
+import bodyParser from 'body-parser'
+import * as sdk from 'botpress/sdk'
+import crypto from 'crypto'
+import { Router } from 'express'
+import _ from 'lodash'
+
+import { Config } from '../config'
+
+const outgoingTypes = ['text', 'typing', 'login_prompt', 'carousel']
+type MessengerAction = 'typing_on' | 'typing_off' | 'mark_seen'
+
+export class MessengerService {
+  private readonly http = axios.create({ baseURL: 'https://graph.facebook.com/v2.6/me' })
+
+  private messengerClients: { [key: string]: MessengerClient } = {}
+  private router: Router
+
+  constructor(private bp: typeof sdk) {}
+
+  initialize() {
+    this.router = this.bp.http.createRouterForBot('channel-messenger', { checkAuthentication: false })
+    this.router.use(
+      bodyParser.json({
+        verify: this._verifySignature.bind(this)
+      })
+    )
+
+    this.router.get('/webhook', this._setupWebhook.bind(this))
+    this.router.post('/webhook', this._handleIncomingMessage.bind(this))
+
+    this.bp.events.registerMiddleware({
+      description: 'Sends outgoing messages for the messenger channel',
+      direction: 'outgoing',
+      handler: this._handleOutgoingEvent.bind(this),
+      name: 'messenger.sendMessages',
+      order: 200
+    })
+  }
+
+  getMessengerClient(botId: string): MessengerClient {
+    if (this.messengerClients[botId]) {
+      return this.messengerClients[botId]
+    }
+
+    this.messengerClients[botId] = new MessengerClient(botId, this.bp, this.http)
+    return this.messengerClients[botId]
+  }
+
+  // See: https://developers.facebook.com/docs/messenger-platform/webhook#security
+  private async _verifySignature(req, res, buffer) {
+    const client = this.getMessengerClient(req.params.botId)
+    const config = await client.getConfig()
+    const signatureError = new Error("Couldn't validate the request signature.")
+
+    if (!/^\/webhook/i.test(req.path)) {
+      return
+    }
+
+    const signature = req.headers['x-hub-signature']
+    if (!signature) {
+      throw signatureError
+    } else {
+      const [, hash] = signature.split('=')
+      const expectedHash = crypto
+        .createHmac('sha1', config.verifyToken)
```

Comment:
Why not use Messenger terminology (accessToken) instead of verifyToken? I was puzzled with that name

Reply:
because thats actually Messenger's terminology https://developers.facebook.com/docs/messenger-platform/getting-started/webhook-setup/